### DISCUSSION TITLE
Trying to produce a VipsImage suitable for Qt6 based on QColorSpace.

### DISCUSSION THREAD

**Discussion by dv1** [libvips/libvips#4282] [Q&A] [answered]
In Qt6, a `QWindow` class has a `format()` function, which returns information about the pixel format and the associated color space that window uses. The color space information is available as a [QColorSpace](https://doc.qt.io/qt-6/qcolorspace.html) instance. It contains information that in theory could be used by libvips to transform a VipsImage to be shown accurately on screen. And this i s my goal - showing VipsImages on screen using Qt6, in an accurate, color space dependent fashion.

However, the properties of QColorSpace do not seem to match what is available in libvips. The closest I found was to use the [iccProfile() function](https://doc.qt.io/qt-6/qcolorspace.html#iccProfile) to produce an ICC profile out of the QColorSpace and then somehow feed this into libvips. But I see no way of supplying an ICC profile from a memory buffer. I could create a temporary file of course, but this seems suboptimal, and I'd prefer a memory buffer.

Am I on the right track here? If so, how can I supply that ICC profile from the QColorSpace? Also, is `vips_icc_transform` the right function to use in this case? Or, can you think of an entirely different, more efficient and appropriate way of accomplishing my goal?

**Comment by jcupitt** [comment, marked as answer]
Hello @dv1,

That sounds interesting. vipsdisp just transforms to sRGB and assumes your display is set to that -- you get the right colour (probably), just with a rather limited gamut.

https://github.com/jcupitt/vipsdisp

So that would be a very simple approach.

You're right, the libvips `icc_transform()` operation should have a way of passing in a dynamically-generated export profile, I'll make an enhancement issue.

Qt's `iccProfile()` function will make a matrix shaper profile, so a 3x3 matrix plus a transfer function, probably encoded as a look-up table. You can just do this processing in libvips with `recomb()` for the 3x3 matrix and `maplut()` for the transfer function, but that would be a more difficult option.

Is this for painting a large image to the screen? You can get libvips to render pixels asynchronously on demand, have a go with vipsdisp and a large image to see it in action. It's on flathub, so it should be very simple to install and try out. There's a minimal example of on-demand rendering here:

https://github.com/libvips/vipsdisp-tiny/blob/master/vipsdisp.c

Just a few 100 lines of C for the whole thing.

**Reply by dv1** [reply]
This is for rendering very large images. I want to do most of the work with VIPS to benefit from its efficient streaming capabilities. Qt6 itself shall only be concerned with the VipsRegion that represents the visible portion, which may be a subset of the image if the UI zooms into the VipsImage. My plan is to create a VipsImage that is of the same size as the Qt6 widget where it is shown, and that uses the same pixel format and same color space / ICC profile.

In total, it would work like this if I understand VIPS correctly:

1. When the image is first loaded, `vips_icc_transform()` is called to convert the original `VipsImage` to one that has its pixels converted to match the output (= the display) ICC profile and the output's pixel format. (The output ICC profile is set to the ICC profile from `QColorSpace::iccProfile()`.) This happens on-demand thanks to the VIPS design. The idea is to do as much of the necessary conversion as possible with VIPS, and produce something that Qt6 can use immediately, with no conversions (or as few as possible).
2. When Qt6 renders the image, a `VipsRegion` is created (if not already existing) that is of the size of the visible part of the image (which may be the entire image, or a subset, if the image is "zoomed in"). `vips_region_prepare()` is used to fill the `VipsRegion` with pixels. On the Qt6 side, a [QImage](https://doc.qt.io/qt-6/qimage.html) is created, with a pixel format that matches what the `VipsRegion` uses. This allows the code to then simply memcpy the pixel rows out of the `VipsRegion` into the `QImage.` The `QImage` is then rendered. (Qt6 takes care of image scaling. I prefer that, since Qt6 can use the GPU for this.)

Note that the `VipsRegion` preparation and `QImage` filling are only done if the zoom or widget size change, or if a different image is loaded. Otherwise, there is no reason for not reusing them.

As far as I can see, this approach would maximize the use of libvips' potential and also ensure that the image is shown in a color accurate fashion.

And the reason why I use the `VipsRegion` as indicated above is that I want to avoid scenarios where I have to wait for the entire image to load before I can do anything with it. That is: Partial image loading is okay. Especially if I zoom into the image, I wish for the zoomed in region to take priority. The rest can be loaded later.

**Reply by dv1** [reply]
To further clarify:

The parts that I want to let VIPS do:

1. Color space conversion using ICC profiles.
2. Color format conversion (like YCbCr -> RGB, or between various pixel formats like 32-bit RGBA and 16-bit RGB if possible).
3. Image loading (very important for very large images).
4. Producing pixels for a region; ideally, this could also be aborted (more on that below).

The parts that I want to let Qt6 do:
1. Scaling of the pixels of the selected `VipsRegion`. Qt6 can efficiently do this in-line when rendering using the GPU.

Now, as for aborting the production of pixels, imagine this situation: Initially, the `VipsRegion` is set to encompass the entire image, that is, the whole image is shown on screen. Loading it takes a lot of time though, so the user might decide to zoom into a subregion before the image is fully loaded. Once that happens, producing the pixels for the entire image needs to be aborted. Instead, VIPS needs to begin producing pixels for the newly zoomed in subregion. From my understanding, this is best accomplished by breaking up the visible part into tiles (for example, 128x128 pixel tiles), and call `vips_region_prepare` on each tile, the tiles being specified as rectangles using `VipsRect`. This then allows for aborting partial loading, since one only has to wait for the ongoing `vips_region_prepare` call on a tile to finish, and then, the tiles from the zoomed in region can be loaded instead. Right? Or perhaps I should use bundles of pixel rows instead of tiles (since some formats like PNG may only load per-scanline)?

**Reply by jcupitt** [reply]
Ah OK, sounds good. 

Yes, I would load as eg. 256x256 tiles, that way when the user pans or zooms about, you can start work on new tiles without much latency.

I would not call `vips_region_prepare()` directly, I would use `vips_sink_screen()` (this is what vipsdisp does).

Using `vips_sink_screen()` would give you much better threading -- it'll run a pool of workers and notify you asynchronously as the workers produce tiles for display (each tile it hands you is a vipsregion, as you said). Tiles are handed back to your main thread, so you don't need to do any locking.

vipsdisp does the scale + composite on the GPU -- it builds a sparse pyramid of tiles using `vips_sink_screen()`, loads them as GPU textures, then scales and composites to the display synchronised with desktop refresh (usually 60fps). It culls GPU tiles against the viewport to avoid loading too much data on there.

It's gtk4 of course, but you could probably chop it about a bit and get it working with Qt. TileSource is the class that makes tiles for display (lazy load, scale, transform to RGB, apply various effects), TileCache maintains the pyramid, the set of tiles on the GPU, and composite for display, Imagedisplay is the image widget, and Imageui adds all the keyboard navigation and zoom animations.

I'd try it out, I spent ages on the stupid thing, it's quite slick. I think it would probably be difficult to make large image display faster or smoother, with gtk4 anyway.

It's the image view component for nip4:

https://opencollective.com/libvips/updates/progress-on-nip4

**Reply by dv1** [reply]
@jcupitt Is there a way to stop a partial image loading process that is running in the background and that began with a `vips_sink_screen` call? When I load a new image in the Qt application, I want to make sure that any processing of the old image is stopped immediately first. I see that internally, the VipsImage uses a "close" signal to disengage rendering, so just unref'ing the image should suffice?

**Reply by jcupitt** [reply]
Yes, you can also set `kill` on the image and it should stop any new threads starting there.

https://www.libvips.org/API/current/VipsImage.html#vips-image-set-kill

**Comment by jcupitt** [comment]
Enhancement issue here: https://github.com/libvips/libvips/issues/4283

**Comment by jcupitt** [comment]
I forgot, there's a slightly crappy youtube video here showing vipsdisp displaying huge images with asynchronous on-demand tile rendering:

https://www.youtube.com/watch?v=KSoBkO2_HtE

Might save you installing it.
